### PR TITLE
fix(compensatory leave request): Create leave ledger entry on submit/cancel

### PR DESCRIPTION
**Pull request by Mangesh-Khairnar**
**Issue:** Ledger entries for compensatory off were not getting created

**Proposed Changes and Additions**
- Create leave ledger entry on submission of each compensatory off request, the period for each such entry would be decided based on the work date field of compensatory off. It will all linked to the same leave allocation that was initially created.
- Create a reverse entry  for every cancellation
- Added test cases to validate the above changes

### CODE REVIEW

File: erpnext/hr/doctype/compensatory_leave_request/compensatory_leave_request.py [hunk 1]
```diff
@@ -68,9 +73,14 @@ def on_cancel(self):
 			leave_allocation = frappe.get_doc("Leave Allocation", self.leave_allocation)
 			if leave_allocation:
 				leave_allocation.new_leaves_allocated -= date_difference
-				if leave_allocation.total_leaves_allocated - date_difference <= 0:
-					leave_allocation.total_leaves_allocated = 0
-				leave_allocation.submit()
+				if leave_allocation.new_leaves_allocated - date_difference <= 0:
+					leave_allocation.new_leaves_allocated = 0
+				leave_allocation.validate()
+				leave_allocation.db_set("new_leaves_allocated", leave_allocation.total_leaves_allocated)
```
Comment (Anurag810) [inline erpnext/hr/doctype/compensatory_leave_request/compensatory_leave_request.py:79]: Why are you setting value (```leave_allocation.db_set```)(when you are creating new leave allocation)? You can pass a flag ```"do_not_submit"``` which will avoid submitting and submit after entering all value. Which will reduce the database interaction for db_set?

**Review by Anurag810 — APPROVED** [review]

**Comment by Anurag810** [comment]
@Mangesh-Khairnar needs to test locally. We will pair and test it locally.

File: erpnext/hr/doctype/compensatory_leave_request/compensatory_leave_request.py [hunk 2]
```diff
@@ -68,9 +73,14 @@ def on_cancel(self):
...
+				leave_allocation.db_set("total_leaves_allocated", leave_allocation.total_leaves_allocated)
+
+				# create reverse entry on cancelation
+				create_additional_leave_ledger_entry(leave_allocation, date_difference * -1, add_days(self.work_end_date, 1))
 
 	def exists_allocation_for_period(self, leave_period):
```
Comment (Anurag810) [inline erpnext/hr/doctype/compensatory_leave_request/compensatory_leave_request.py]: Change function name like ```get_existing_allocation_for_period```

File: erpnext/hr/doctype/compensatory_leave_request/compensatory_leave_request.py [hunk 3]
```diff
@@ -53,7 +52,13 @@ def on_submit(self):
...
+				leave_allocation.validate()
+				leave_allocation.db_set("new_leaves_allocated", leave_allocation.total_leaves_allocated)
+				leave_allocation.db_set("total_leaves_allocated", leave_allocation.total_leaves_allocated)
+
+				# generate additional ledger entry for the new compensatory leaves off
+				create_additional_leave_ledger_entry(leave_allocation, date_difference, add_days(self.work_end_date, 1))
+
 			else:
 				leave_allocation = self.create_leave_allocation(leave_period, date_difference)
 			self.db_set("leave_allocation", leave_allocation.name)
```
Comment (Anurag810) [inline erpnext/hr/doctype/compensatory_leave_request/compensatory_leave_request.py]: Check indentation and whether ```db_set ``` is required or not?

File: erpnext/hr/doctype/compensatory_leave_request/test_compensatory_leave_request.py [hunk 4]
```diff
@@ -5,37 +5,101 @@
...
+		frappe.db.sql(''' delete from `tabLeave Ledger Entry`''')
+		frappe.db.sql(''' delete from `tabLeave Allocation`''')
+		frappe.db.sql(''' delete from `tabAttendance` where attendance_date=%s ''', (today()))
+		create_leave_period(add_months(today(), -3), add_months(today(), 3), "_Test Company")
+		create_holiday_list()
+
+		employee = get_employee()
+		employee.holiday_list = "_Test Compensatory Leave"
+		employee.save()
+
```
Comment (Anurag810) [inline erpnext/hr/doctype/compensatory_leave_request/test_compensatory_leave_request.py:25]: Write the test for update and creation of leave allocation on CLS submission and cancellation

**Review by Anurag810 — CHANGES_REQUESTED** [review]

File: erpnext/hr/doctype/compensatory_leave_request/test_compensatory_leave_request.py (hunk 4 above)
Comment (Mangesh-Khairnar) [reply erpnext/hr/doctype/compensatory_leave_request/test_compensatory_leave_request.py:25]: added test

File: erpnext/hr/doctype/compensatory_leave_request/compensatory_leave_request.py (hunk 3 above)
Comment (Mangesh-Khairnar) [reply erpnext/hr/doctype/compensatory_leave_request/compensatory_leave_request.py]: removed db_set

File: erpnext/hr/doctype/compensatory_leave_request/compensatory_leave_request.py (hunk 2 above)
Comment (Mangesh-Khairnar) [reply erpnext/hr/doctype/compensatory_leave_request/compensatory_leave_request.py]: done

**Pull request merged**